File: src/cli/exit.ts
/**
 * Exit-code policy (E5 fitness function: matrix stays stable).
 *
 *   0 — success
 *   1 — user error: bad argv, drift abort, non-TTY swap with no --on-drift,
 *       validation failure, missing profile name on the CLI (typo — the user
 *       can fix it by editing their invocation)
 *   2 — system error: unexpected IO/permission/internal fault, including
 *       not-yet-implemented stubs (init/hook in E5)
 *   3 — conflict / cycle / structural-missing: ResolverError that signals
 *       the manifest graph itself is broken (cycle, missing profile reached
 *       via an extends chain, missing include); LockHeldError also maps here
 *       (a peer process holds the project, conceptually a "missing slot")
 *
 * The 'missing profile' case is split: when the user types
 * `claude-profiles use ghst` (typo for `ghost`), the resolver throws
 * MissingProfileError with referencedBy === undefined — that's a CLI typo
 * and exits 1. When a manifest declares `extends: "nope"`, the resolver
 * throws the same error class but with referencedBy set — that's a
 * structural manifest failure and exits 3.
 *
 * Mapping is centralised so individual command handlers don't reinvent it.
 */

import { LockHeldError } from "../state/lock.js";

import {
  ConflictError,
  CycleError,
  MissingIncludeError,
  MissingProfileError,
  MergeError,
  PipelineError,
  ResolverError,
} from "../errors/index.js";

export const EXIT_OK = 0;
export const EXIT_USER_ERROR = 1;
export const EXIT_SYSTEM_ERROR = 2;
export const EXIT_CONFLICT = 3;

/** Fixed CLI exit codes — the test suite asserts each maps to the right value. */
export type ExitCode =
  | typeof EXIT_OK
  | typeof EXIT_USER_ERROR
  | typeof EXIT_SYSTEM_ERROR
  | typeof EXIT_CONFLICT;

/**
 * Marker thrown by command handlers to signal a user-error exit without a
 * specific error subclass (e.g. drift abort, validation fail). Carries the
 * intended exit code so the dispatcher doesn't have to guess.
 */
export class CliUserError extends Error {
  readonly exitCode: ExitCode;
  constructor(message: string, exitCode: ExitCode = EXIT_USER_ERROR) {
    super(message);
    this.name = "CliUserError";
    this.exitCode = exitCode;
  }
}

/**
 * Marker for "not yet implemented in this epic" — used by E5 stubs for the
 * init/hook verbs (E6 owns those). Maps to exit 2 (system error class) so
 * callers don't conflate it with bad user input.
 */
export class CliNotImplementedError extends Error {
  constructor(verb: string, owner: string) {
    super(`${verb} is not yet implemented (owned by ${owner})`);
    this.name = "CliNotImplementedError";
  }
}

/**
 * Map any error thrown by the command pipeline to a CLI exit code.
 *
 * - CliUserError carries its own code (lets callers force exit 1 vs 3).
 * - MissingProfileError → 1 if it's a CLI typo (referencedBy undefined),
 *   3 if it's structural (manifest extends chain points at nothing).
 * - Other ResolverError subclasses → 3 (conflict/cycle/missing-include).
 * - LockHeldError → 3 (the project is occupied; semantically "no slot for us").
 * - MergeError → 2 (runtime drift, not user-input fault).
 * - CliNotImplementedError → 2 (system error class).
 * - Anything else → 2 (unexpected system fault).
 */
export function exitCodeFor(err: unknown): ExitCode {
  if (err instanceof CliUserError) return err.exitCode;
  if (err instanceof CliNotImplementedError) return EXIT_SYSTEM_ERROR;
  if (err instanceof LockHeldError) return EXIT_CONFLICT;
  if (err instanceof MissingProfileError) {
    // Distinguish "user typed a name that doesn't exist" (typo, fixable by
    // editing argv → exit 1) from "manifest's extends chain points at a
    // profile that doesn't exist" (structural fault → exit 3). The resolver
    // sets referencedBy only when the missing name was reached transitively.
    return err.referencedBy === undefined ? EXIT_USER_ERROR : EXIT_CONFLICT;
  }
  if (
    err instanceof ConflictError ||
    err instanceof CycleError ||
    err instanceof MissingIncludeError
  ) {
    return EXIT_CONFLICT;
  }
  if (err instanceof ResolverError) return EXIT_CONFLICT;
  if (err instanceof MergeError) return EXIT_SYSTEM_ERROR;
  if (err instanceof PipelineError) return EXIT_SYSTEM_ERROR;
  return EXIT_SYSTEM_ERROR;
}

File: internal/cli/exit.go
// Exit-code policy. Mirrors src/cli/exit.ts. The matrix is frozen — new
// error variants must route through one of the four codes via ExitCodeFor.
package cli

import (
	"errors"

	"github.com/htxryan/claude-code-config-profiles/internal/cli/commands"
	"github.com/htxryan/claude-code-config-profiles/internal/cli/service"
	pipelineerrors "github.com/htxryan/claude-code-config-profiles/internal/errors"
	"github.com/htxryan/claude-code-config-profiles/internal/state"
)

// Exit codes are the public CLI surface (E5 fitness function).
const (
	ExitOK       = 0
	ExitUser     = 1
	ExitSystem   = 2
	ExitConflict = 3
)

// CliUserError is a marker thrown by command handlers to signal a user-error
// exit without a specific error subclass (drift abort, validation fail).
type CliUserError struct {
	Message  string
	ExitCode int
}

func (e *CliUserError) Error() string { return e.Message }

// NewUserError builds a *CliUserError at the user-error code (1).
func NewUserError(msg string) *CliUserError {
	return &CliUserError{Message: msg, ExitCode: ExitUser}
}

// NewConflictError builds a *CliUserError at the conflict code (3). Used
// when the failure is structural rather than typo-class.
func NewConflictError(msg string) *CliUserError {
	return &CliUserError{Message: msg, ExitCode: ExitConflict}
}

// CliNotImplementedError marks "not yet implemented in this epic" stubs.
// Maps to exit 2 (system error class) — distinct from bad user input.
type CliNotImplementedError struct {
	Verb  string
	Owner string
}

func (e *CliNotImplementedError) Error() string {
	return e.Verb + " is not yet implemented (owned by " + e.Owner + ")"
}

// ExitCodeFor maps any error returned by the command pipeline to a CLI
// exit code. The mapping mirrors src/cli/exit.ts.
func ExitCodeFor(err error) int {
	if err == nil {
		return ExitOK
	}
	// CliUserError carries its own code (lets callers force exit 1 vs 3).
	var cue *CliUserError
	if errors.As(err, &cue) {
		return cue.ExitCode
	}
	// Swap-orchestrator abort signal — mirrors TS CliUserError default.
	if service.IsSwapAbort(err) {
		return ExitUser
	}
	// Per-command user error (commands package can't import cli/).
	var ue *commands.UserError
	if errors.As(err, &ue) {
		if ue.ExitCode != 0 {
			return ue.ExitCode
		}
		return ExitUser
	}
	var nie *CliNotImplementedError
	if errors.As(err, &nie) {
		return ExitSystem
	}
	// Lock-held: a peer process holds the project, conceptually "no slot".
	var lhe *state.LockHeldError
	if errors.As(err, &lhe) {
		return ExitConflict
	}
	// Schema-too-new: refuse to operate. Conflict class — the user must
	// upgrade c3p, not edit input.
	var stne *state.SchemaTooNewError
	if errors.As(err, &stne) {
		return ExitConflict
	}
	// Pipeline errors: split by code so the matrix matches TS exit.ts.
	if pe := pipelineerrors.AsPipelineError(err); pe != nil {
		switch pe.ErrorCode() {
		case pipelineerrors.CodeMissingProfile:
			// Top-level CLI typo (referencedBy=="") → 1; structural manifest
			// failure → 3. Inspect the concrete error to discriminate.
			var mpe *pipelineerrors.MissingProfileError
			if errors.As(err, &mpe) {
				if mpe.ReferencedBy == "" {
					return ExitUser
				}
				return ExitConflict
			}
			return ExitConflict
		case pipelineerrors.CodeCycle,
			pipelineerrors.CodeMissingInclude,
			pipelineerrors.CodeConflict,
			pipelineerrors.CodePathTraversal:
			return ExitConflict
		case pipelineerrors.CodeInvalidManifest:
			// Structural manifest fault — user remediation, but exits 1 to
			// match TS validate semantics (a single bad profile.json doesn't
			// shut down the whole pipeline at conflict-class).
			return ExitUser
		case pipelineerrors.CodeRootMarkers:
			// Materialize-phase markers-missing — user runs `c3p init` to fix,
			// so it's user-error class (TS exit.ts MaterializeError → 1).
			return ExitUser
		case pipelineerrors.CodeInvalidSettings,
			pipelineerrors.CodeMergeReadFailed:
			// Merge-phase faults — runtime drift, not config typo.
			return ExitSystem
		default:
			return ExitSystem
		}
	}
	return ExitSystem
}
